### PR TITLE
Render arabic rtl text properly

### PR DESCRIPTION
**Pull request by solomancode**
## Main Components:
Tokenizer: convert a string into a list of tokens
FeatureQuery: query a feature and lookup a substitution
Bidi: query, apply arabic features and adjust arabic layout
Char: character class assertions

## Supported GSUB Tables
Single Substitution - lookup type 1, substFormat 2
Chaining Context Substitution - lookup type 6, substFormat 3
Ligature Substitution - lookup type 4, substFormat 1

## Types of changes
  New feature (non-breaking change which adds functionality)

Overview

Checklist:
- [x] I did npm run test and all tests passed green (including code styling checks).
- [x] I have added tests to cover my changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the README accordingly.
- [x]  I have read the CONTRIBUTING document.

### CODE REVIEW

**Comment by solomancode** [comment]
Check this:
[https://drive.google.com/file/d/151gmIwggxcMtRuOh3cHk11MFo3_b-Y6V/view?usp=sharing](url)

File: src/features/featureQuery.js [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+        if (substitution.length > 0) {
+            substitutions.push({ id: 41, substitution });
+        }
+    }
+    return substitutions;
+};
+
+/**
+ * Lookup a coverage table
+ * @param {any} contextParams context params to lookup
```
Comment (Jolg42) [inline src/features/featureQuery.js]: Maybe replace that line by
 * @param {number} glyphIndex to lookup

File: src/bidi.js [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+ * @param {string} text input text
+ */
+Bidi.prototype.getBidiText = function (text) {
+    this.processText(text);
+    return this.tokenizer.getText();
+};
+
+/**
+ * Get the current state index of each token
+ * @param {text} text an input text
```
Comment (Jolg42) [inline src/bidi.js]: replace `text` by `string`?

File: src/char.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
+// ╭─┄┄┄────────────────────────┄─────────────────────────────────────────────╮
+// ┊ Character Class Assertions ┊ Checks if a char belongs to a certain class ┊
+// ╰─╾──────────────────────────┄─────────────────────────────────────────────╯
+// jscs:disable maximumLineLength
+/**
+ * Check if a char is Arabic
+ * @param {char} c a single char
```
Comment (Jolg42) [inline src/char.js]: replace all `{char}...` by `{string} a single char`?

File: src/font.js [hunk 4]
```diff
@@ -8,9 +8,11 @@ import Position from './position';
 import Substitution from './substitution';
 import { isBrowser, checkArgument, arrayBufferToNodeBuffer } from './util';
 import HintingTrueType from './hintingtt';
+import Bidi from './bidi';
+import FeatureQuery from './features/featureQuery';
 
 // This code is based on Array.from implementation for strings in https://github.com/mathiasbynens/Array.from
```
Comment (Jolg42) [inline src/font.js]: Remove unused code?

File: src/font.js [hunk 5]
```diff
@@ -8,9 +8,11 @@ import Position from './position';
...
-const arrayFromString = Array.from || (s => s.match(/[\uD800-\uDBFF][\uDC00-\uDFFF]?|[^\uD800-\uDFFF]|./g) || []);
+// const arrayFromString = Array.from || (s => s.match(/[\uD800-\uDBFF][\uDC00-\uDFFF]?|[^\uD800-\uDFFF]|./g) || []);
```
Comment (Jolg42) [inline src/font.js]: Remove these 2 commemted lines?

**Comment by solomancode** [comment]
Hi @Jolg42 ,

Thanks for your comment :relaxed:

As for my implementation, No not at all. It's not inspired by another library. I made a couple of refactors before reaching this stage. I can tell you more about it if you're interested to know more.

**The state of this implementation:**

| Feature | Severity | ref |
| ------------- | ------------- | ------------- |
| rendering of Arabic letter positional forms ( Initial, Medial, Final and Independent )  | **4** | [**?**](https://w3c.github.io/alreq/#h_joining) |
| rendering of Arabic required letter ligature | **4** | [**?**](https://w3c.github.io/alreq/#figure-laam-alef-ligature) |
| Bidirectional text - correct order for Arabic words and sentences | **4** | [**?**](https://w3c.github.io/alreq/#figure-order_when_displayed) |
| Partial support for Arabic Tashkeel (diacritics) ligatures | **3** | [**?**](https://w3c.github.io/alreq/#h_diacritics) |

_**One issues** I found related to Arabic Tashkeel ligatures_
tashkeel_issue

Some Tashkeel characters aren't **composed** correctly. _highlighted in Blue_
Some Tashkeel characters aren't **positioned** correctly. _highlighted in Green

**Comment by solomancode** [comment]
Sorry I accidentally closed the PR. :blush:

**Comment by Jolg42** [comment]
Great!

Just copy-pasting the severity scale that you used as a reference:
```
0 = I don't agree that this is a usability problem at all
1 = Cosmetic problem only: need not be fixed unless extra time is available on project
2 = Minor usability problem: fixing this should be given low priority
3 = Major usability problem: important to fix, so should be given high priority
4 = Usability catastrophe: imperative to fix this before product can be released
```

@solomancode As you rated most of the features 4 or 3 I guess we can't say it's a full support for sure but maybe something like "Partial support of arabic text rendering (For more details see [#359](https://github.com/nodebox/opentype.js/pull/359))" in the README?

Also you rated some features 4 but they are "working" I think? At least partially? (Nothing crashed when I tested your PR on my side)
Are these features good enough to be rated 3 maybe? The level 4 `Usability catastrophe` sounds scary to me! It sounds like it doesn't work at all. If you have more details on why it's a level 4 it would be great 😃

**Comment by solomancode** [comment]
Sorry, My bad. :smile: . I should've explained it more clearly. 

I think you're confusing **progress** with **severity**.

All the features in table 1 are complete and resolves an issue of a certain severity. I'll give you an example

**Feature 1:**
_rendering of Arabic letter positional forms ( Initial, Medial, Final and Independent )_

without this feature Arabic letters will be rendered separately in a single form.

- _Below you can see my name '**Mohamed**' before applying positional forms. As you can see letters are not joined together and all of them are rendered in the original form. The issue here is that it's not a word anymore, It'll be read as a sequence of letters and that is why I evaluated this issue as a **level 4** ( we can't say that we're supporting Arabic without fixing this issue )_

muhamed_before

- _After applying positional form. as you can see depending on the position of each letter and the state of it's preceding letter ( isolated or not ). A letter can take some form — It's the same letter but in a different form and the reader will recognize it as a part of a word not a separated letter_

muhamed_after

Another example is the last feature from table 1
_Partial support for Arabic Tashkeel (diacritics) ligatures_

This feature isn't fully complete — partial support. It is an important feature to include but as I mentioned before, Arabic text can be read without Tashkeel (diacritics). an Arabic reader can read can assume the Tashkeel marks without them existing. of course a different Tashkeel for the same word may give an entirely different meaning, but depending on the context a reader can accurately infer which meaning is meant to be convoyed. I evaluated this feature as resolving feature of **level 3** because I think at this stage no one will complain that we're not supporting Tashkeel. It's an important feature to add and I'll be working on it as soon as possible. But at this stage we're good.

**one last note**, the second table is a **TODO** table it's progress is **0%** but as you can see it's quite severe.

**Comment by zoghal** [comment]
i like it this feature

**Comment by solomancode** [comment]
I forgot to mention the **state of supporting Arabic Tashkeel** (diacritics).

- we're supporting rendering of isolated Tashkeel marks but we're having an issue with **composing multiple marks** _( I'll see to it later )_.
- mark position is a little off, I think the solution to this issue resides in the **GPOS** table.

**Comment by solomancode** [comment]
Hi @Jolg42,

I fixed the Tashkeel issue, I forgot to reset the rendering base direction to the default rendering direction 'ltr'.

compo

**Comment by Jolg42** [comment]
Nice!

I'll add the mention of this in the README and publish it asap 😉

*For reference, I found more details about Tashkeel at https://en.wikipedia.org/wiki/Arabic_diacritics*

**Comment by Jolg42** [comment]
Feature published with [0.11.0](https://github.com/opentypejs/opentype.js/tree/0.11.0) 🚀

**Pull request merged**